lib/Language/Souffle/Interpreter.hs

```diff
@@ -0,0 +1,237 @@
+    handle <- readIORef ref
+    let srcPath = factPath handle
+    copyFiles srcPath "./facts"
+
+  -- | Returns all facts of a program. This function makes use of type inference
+  --   to select the type of fact to return.
+  getFacts :: forall a prog . (Marshal a, Fact a, ContainsFact prog a)
+           => Handle prog -> SouffleM [a]
+  getFacts (Handle ref) = SouffleM $ liftIO $do
+    handle <- readIORef ref
+    let relationName = factName (Proxy :: Proxy a)
+    let factFile = outputPath handle </> relationName <.> "csv"
+    factLines <- readCSVFile factFile
+    let facts = map (runMarshallIT pop) factLines
+    pure facts
+
+  -- | Searches for a fact in a program.
+  --   Returns 'Nothing' if no matching fact was found; otherwise 'Just' the fact.
+  --
+  --   Conceptually equivalent to @List.find (== fact) \<$\> getFacts prog@, but this operation
+  --   can be implemented much faster.
+  findFact :: (Fact a, ContainsFact prog a, Eq a, Marshal a)
+           => Handle prog -> a -> SouffleM (Maybe a)
+  findFact prog fact = find (== fact) <$> getFacts prog
+
+  -- | Adds a fact to the program.
+  addFact :: forall a prog . (Fact a, ContainsFact prog a, Marshal a)
+          => Handle prog -> a -> SouffleM ()
+  addFact (Handle ref) fact = SouffleM $ do
+    handle <- readIORef ref
+    let relationName = factName (Proxy :: Proxy a)
+    let factFile = factPath handle </> relationName <.> "csv"
+    let line = pushMarshallIT (push fact)
+    appendFile factFile $ intercalate "\t" line ++ "\n"
+
+  -- | Adds multiple facts to the program. This function could be implemented
+  --   in terms of 'addFact', but this is done as a minor optimization.
+  addFacts :: forall a prog f . (Fact a, ContainsFact prog a, Marshal a, Foldable f)
+           => Handle prog -> f a -> SouffleM ()
+  addFacts (Handle ref) facts = SouffleM $ do
+    handle <- readIORef ref
+    let relationName = factName (Proxy :: Proxy a)
+    let factFile = factPath handle </> relationName <.> "csv"
+    let factLines :: [[String]] = map (pushMarshallIT . push) (foldMap pure facts)
+    mapM_ (\line -> appendFile factFile (intercalate "\t" line ++ "\n")) factLines
+
+copyFiles :: FilePath -> FilePath -> IO ()
+copyFiles srcPath destPath = do
+  content <- listDirectory srcPath
+  forM_ content $ \file -> do
+    doesFileExist (srcPath </> file) >>= \case
+      False -> pure () -- Dir or symlink, but not file
+      True  -> copyFile (srcPath </> file) (destPath </> file)
+
+datalogProgramFile :: forall prog . Program prog => prog -> IO FilePath
+datalogProgramFile _prg = do
+  dir <- fromMaybe "." <$> lookupEnv "DATALOG_DIR"
+  pure $ dir </> programName (Proxy :: Proxy prog) <.> "dl"
+
+souffleCmd :: IO FilePath
+souffleCmd = fromMaybe "souffle" <$> lookupEnv "SOUFFLE_BIN"
+
+readCSVFile :: FilePath -> IO [[String]]
+readCSVFile = fmap (fmap (splitOn "\t") . lines) . readFile
+
+cleanUp :: forall prog . (Program prog) => Handle prog -> SouffleM ()
```

Comment:
cleanUp reallyb interpreter specific. I left it out from the run or the MonadSouffle typeclass.

Reply:
We can add this to `runSouffle` for interpreter version.